content/posts/2015-09-02-heroku-build-static.md

```diff
@@ -0,0 +1,119 @@
+have to maintain them. Not maintaining infrastructure like this is the reason I
+use Heroku, so this seems like one step forward, one step back.
+
+[search]: https://github.com/search?utf8=%E2%9C%93&q=heroku+buildpack+python+node&type=Repositories&ref=searchresults
+
+Enter [Multi Buildpack][], which combines multiple Heroku buildpacks in a
+pretty sane way. It is a very 'thin' buildpack, so there isn't much to fall out
+of compatibility. I also assume that because Heroku has a fork of it, it will
+stay up to date anyways. It combines buildpacks, so it can combine the official
+Python and Node buildpacks. That means that it probably won't break randomly in
+the future.
+
+[Multi Buildpack]: https://github.com/heroku/heroku-buildpack-multi
+
+To configure the buildpack, first tell Heroku you want to use it
+
+```bash
+$ heroku buildpacks:set https://github.com/heroku/heroku-buildpack-multi.git
+```
+
+And then add a `.buildpacks` file to your repository, so the multi buildpack knows
+what to do.
+
+<span class="codepath">./.buildpacks</span>
+
+```text
+https://github.com/heroku/heroku-buildpack-nodejs.git
+https://github.com/heroku/heroku-buildpack-python.git
+```
+
+The order here is the order that the buildpacks will run in, so it can be
+significant. In theory, earlier buildpacks will be available in later build
+packs, which can be useful for build tools.
+
+Actually running it all
+=======================
+
+Unfortunately, in practice this doesn't totally work. Both buildpacks run. Both
+Python and Node environments are available at runtime. However, due to hilarity
+involving symlinks, relative paths, and moving files, the Python buildpack works
+in a way that makes the Node environment unavailable.
+
+The Python buildpack automatically detects a Django project and runs
+`./manage.py collectstatic`. This is a pretty neat trick, that makes most
+Django development work nicely. But the Node environment isn't available, so
+this fails, and we are back to square one. No static files get built.
+
+The solution lies in `bin/post_compile`. If there is an executable at that
+path, Heroku will run it at the end of the build process. Since it runs after
+the Python buildpack has undone the silly path tricks, the Node environment
+will work. It is still required to fiddle with PATHs a little to make it all
+work, but work it does!
+
+This trick will work with any Python webapp. In the Django projects I do this
+trick with, we use [Django Pipeline][]. All the static asset compilation
+happens through a `manage.py collectstatic` command. This calls out to the Node
+tools, so the build step relies on Node. Calling Node tools like Gulp or
+Webpack would also work nicely, if that's how you roll.
+
+[Django Pipeline]: https://github.com/cyberdelia/django-pipeline
+
+<span class="codepath">./bin/post_compile</span>
+
+```bash
+#!/bin/bash
+export PATH=/app/.heroku/node/bin:$PATH
+./manage.py collectstatic --noinput
+```
+
```

Comment:
Maybe an extra paragraph here addressing the path hack:

> (Okay, so maybe we had to be a little tricky by appending the node binary folder to the `PATH`. Pretend you didn't see that! Or don't, because you'll need to do it in your script too.)